src/ray/raylet/worker_pool.h

```diff
@@ -282,13 +288,13 @@ class WorkerPool : public WorkerPoolInterface, public IOWorkerPoolInterface {
   /// the worker back onto the pool once the worker has completed its work.
   ///
   /// \param task_spec The returned worker must be able to execute this task.
-  /// \param allocated_instances_serialized_json The allocated resouce instances
-  /// json string.
-  /// \return An idle worker with the requested task spec. Returns nullptr if no
-  /// such worker exists.
-  std::shared_ptr<WorkerInterface> PopWorker(
-      const TaskSpecification &task_spec,
-      const std::string &allocated_instances_serialized_json = "{}");
+  /// \param callback The callback function that executed when gets the result of
+  /// worker popping.
```

Comment:
I am a bit confused about the possible triggers for the callback. Can you comment and say what they are? In particular, I'm not sure about what happens when:
1. There's no job config available yet.
2. There's no available worker (there seems to be multiple cases here?)

I might be wrong, but it also doesn't seem like the unit tests in worker_pool_tests check whether we return the right error codes and at the right time in these cases. Can we add tests for these?

Reply:
If there's no job config available or the starting workers number greater than maximum_startup_concurrency_, we will callback an empty worker to trigger re-dispatch. And If runtime env creation failed, we should also callback empty worker but trigger re-schedule to other nodes or mark task dead. But I haven't implemented this in current PR, just add some status codes.

Reply:
Thanks, can we make sure to document what the cases are where will trigger the callback? I noticed you added status codes, but it would be good to explain what they mean in the comments.